File: Sample.java
// Delete and Earn

// Time Complexity : O(n)
// Space Complexity : O(n)

class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        if not nums:
            return 0
        
        max_num = max(nums)

        points = [0] * (max_num+1)
        for num in nums:
            points[num] += num

        dp = [0] * (max_num+1)
        dp[1] = points[1]

        for i in range(2,max_num+1):
            dp[i] = max(dp[i-1], dp[i-2]+points[i])

        return dp[max_num]

//  Minimum Falling Path Sum
// Time Complexity : O(mn)
// Space Complexity : O(1)

class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        m = len(matrix)
        n = len(matrix[0])
        min_path_sum = 0
        if m==0: return 0
        for i in range(1,m):
            for j in range(n):
                aboveSum = matrix[i-1][j]
                aboveRSum = matrix[i-1][j+1] if(j<n-1) else float(inf)
                aboveLSum = matrix[i-1][j-1] if(j>0) else float(inf)
                matrix[i][j] += min(aboveSum, aboveRSum, aboveLSum)
        return min(matrix[m-1])



  

File: Sample.py
# Problem 1: Delete and Earn -> Time Complexicity: O(max(max(nums),n)), Space Complexicity: O(max(nums))

class Solution:
    
    def deleteAndEarn(self, nums: List[int]) -> int:
    
        """
        Convert the problem to ordinary house robber problem by:
        - placing all the sum of a unique value represented by index of the resulting array
        """
        if len(nums) == 0: return 0
        maximum = 0;
        maximum = max(nums)
        aux_array = [0]*(maximum+1)
        
        for i in nums:
            aux_array[i] += i
            
        skip = 0; take = aux_array[0]
        for i in range(len(aux_array)):
            temp = skip
            skip = max(skip, take)
            take = temp + aux_array[i]
        
        return max(skip, take)

# Problem 2: Minimum falling sum for square matrix -> Time Complexicity: O(n^2), Space Complexicity: O(n^2)

class Solution:
    
    def minFallingPathSum(self, A: List[List[int]]) -> int:
        
        """
        Solve by creating an square dp array:
        - Start from bottom row of matrix A
        - add minimum values to the sum by using the constraints
        """
        
        if len(A) == 0: return 0
        
        n = len(A)
        dp = [[0 for __ in range(n)] for __ in range(n)] # Initialize
        dp[n-1] = A[n-1] # Start
        
        for row in reversed(range(n-1)):
            for i in range(n):
                if i != 0 and i != n-1:
                    dp[row][i] = min(dp[row+1][i-1], dp[row+1][i], dp[row+1][i+1]) + A[row][i]
                elif i == 0:
                    dp[row][i] = min(dp[row+1][i], dp[row+1][i+1]) + A[row][i]
                elif i == n-1:
                    dp[row][i] = min(dp[row+1][i-1], dp[row+1][i]) + A[row][i]
        
        return min(dp[0])
